src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -0,0 +1,2814 @@
+}
+";
+            var comp = CreateStandardCompilation(source, references: s_valueTupleRefs, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "Success");
+        }
+
+        [Fact]
+        public void TestEqualOnNullableVsNullableTuples_WithImplicitConversion()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        Compare(null, null);
+        Compare(null, (1, 2));
+        Compare((2, 3), null);
+        Compare((4, 4), (4, 4));
+        Compare((5, 5), (10, 10));
+    }
+    private static void Compare((int, int)? nt1, (byte, long)? nt2)
+    {
+        System.Console.Write($""{nt1 == nt2} "");
+    }
+}
+";
+            var comp = CreateStandardCompilation(source, references: s_valueTupleRefs, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "True False False True False");
+
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+
+            var comparison = tree.GetCompilationUnitRoot().DescendantNodes().OfType<BinaryExpressionSyntax>().Single();
+            var nt1 = comparison.Left;
+            var nt1Type = model.GetTypeInfo(nt1);
+            Assert.Equal("nt1", nt1.ToString());
+            Assert.Equal("(System.Int32, System.Int32)?", nt1Type.Type.ToTestDisplayString());
+            Assert.Equal("(System.Int32, System.Int64)?", nt1Type.ConvertedType.ToTestDisplayString());
+
+            var nt2 = comparison.Right;
+            var nt2Type = model.GetTypeInfo(nt2);
+            Assert.Equal("nt2", nt2.ToString());
+            Assert.Equal("(System.Byte, System.Int64)?", nt2Type.Type.ToTestDisplayString());
+            Assert.Equal("(System.Int32, System.Int64)?", nt2Type.ConvertedType.ToTestDisplayString());
+
+            verifier.VerifyIL("C.Compare", @"{
+  // Code size      217 (0xd9)
+  .maxstack  3
+  .locals init ((int, long)? V_0,
+                bool V_1,
+                System.ValueTuple<int, long> V_2,
+                (int, long)? V_3,
+                System.ValueTuple<int, long> V_4,
+                (int, int)? V_5,
+                (int, long)? V_6,
+                System.ValueTuple<int, int> V_7,
+                (byte, long)? V_8,
+                System.ValueTuple<byte, long> V_9)
+  IL_0000:  nop
+  IL_0001:  ldstr      ""{0} ""
+  IL_0006:  ldarg.0
+  IL_0007:  stloc.s    V_5
+  IL_0009:  ldloca.s   V_5
+  IL_000b:  call       ""bool (int, int)?.HasValue.get""
+  IL_0010:  brtrue.s   IL_001e
+  IL_0012:  ldloca.s   V_6
+  IL_0014:  initobj    ""(int, long)?""
+  IL_001a:  ldloc.s    V_6
+  IL_001c:  br.s       IL_0040
+  IL_001e:  ldloca.s   V_5
+  IL_0020:  call       ""(int, int) (int, int)?.GetValueOrDefault()""
+  IL_0025:  stloc.s    V_7
+  IL_0027:  ldloc.s    V_7
+  IL_0029:  ldfld      ""int System.ValueTuple<int, int>.Item1""
+  IL_002e
```

Comment:
Are we testing that tuple expressions are executed exactly once?
```
b = HasSideEffects() == (1, 2);
b = (1, HasSideEffects()) == (1, (2, 3));
```

Reply:
`TestEvaluationOrderOnTupleType` covers the second case. I'll add a test for the first one.


---
In reply to: [167992038](https://github.com/dotnet/roslyn/pull/23356#discussion_r167992038) [](ancestors = 167992038)